### PR TITLE
kernel: init: allow to jitter boot delay

### PR DESCRIPTION
**Pull request by markus-becker-tridonic-com**
When a big amount of networked devices  collectively gets switched on,
e.g. by a mains switch, quickly overwhelms a wireless network.

This PR allows to jitter the start-up of the devices.

Cleaned-up the boot banner after the boot delay. At that point in time
is irrelevant.

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

### CODE REVIEW

**Comment by markus-becker-tridonic-com** [comment]
Open issue: How to make sure that the printk() is fllushed before k_msleep()?

**Comment by pfalcon** [comment]
> When a big amount of networked devices collectively gets switched on,
e.g. by a mains switch, quickly overwhelms a wireless network.

As this mentions networking usecase, adding @jukkar to reviewers. I myself would point that if networking is *the* usecase, than rather e.g. DHCP client should have a randomized delay support. Handling that with a boot delay seems like a crude approach.

**Comment by markus-becker-tridonic-com** [comment]
It is mostly a wireless/BT/openThread networking use case, not all of the networking options necessarily require DHCP. 

It might be a crude approach, but the implementation is rather straight forward.

File: kernel/Kconfig [hunk 1]
```diff
@@ -327,6 +327,14 @@ config BOOT_DELAY
 	  achieved by waiting for DCD on the serial port--however, not
 	  all serial ports have DCD.
 
+config BOOT_DELAY_RANDOM
+	bool "Randomize boot delay"
+	default n
```
Comment (jukkar) [inline kernel/Kconfig:332]: No need to have `default n` as the option is disabled by default.

File: kernel/Kconfig [hunk 2]
```diff
@@ -327,6 +327,14 @@ config BOOT_DELAY
...
+	help
+	  This option randomizes the bootup delay between 0 and the
+	  value of BOOT_DELAY. This is used to jitter boot-up of
```
Comment (jukkar) [inline kernel/Kconfig:335]: Instead of
`This is used to jitter boot-up of wirelessly networked devices.`
I would say
`This can be used e.g., to jitter boot-up of wireless network devices.`
to emphasize that this is just one use case.

File: kernel/init.c [hunk 3]
```diff
@@ -219,18 +218,17 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
 	z_stack_adjust_initialized = 1;
 #endif
 	if (boot_delay > 0 && IS_ENABLED(CONFIG_MULTITHREADING)) {
-		printk("***** delaying boot " STRINGIFY(CONFIG_BOOT_DELAY)
-		       "ms (per build configuration) *****\n");
-		k_busy_wait(CONFIG_BOOT_DELAY * USEC_PER_MSEC);
+		printk("***** Delaying boot %u ms *****\n", boot_delay);
+		k_msleep(boot_delay);
```
Comment (jukkar) [inline kernel/init.c:222]: Shouldn't we use `k_busy_wait()` here like in the original version?

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks reasonable. And as already commented, the DHCP provides similar support but if one is not using DHCP then something like this is useful.

File: kernel/init.c (hunk 3 above)
Comment (markus-becker-tridonic-com) [reply kernel/init.c:222]: With `k_msleep()` I could get the `printk()` message on UART before the boot_delay elapsed, with `k_busy_wait()` only after the delay elapsed and by then it is not useful anymore (me as a user was wondering what is happening).

File: kernel/Kconfig (hunk 1 above)
Comment (markus-becker-tridonic-com) [reply kernel/Kconfig]: done

File: kernel/Kconfig (hunk 2 above)
Comment (markus-becker-tridonic-com) [reply kernel/Kconfig]: done

File: kernel/init.c (hunk 3 above)
Comment (andyross) [reply kernel/init.c:222]: What platform?   That sounds like your logging is setup async, and doing a sleep allows a thread to run to do the output.  If so... that's not a bug, it's what you asked for, technically.  In general, printk() isn't synchronous any more (and yes, I hate this too).

Sleeping in the main thread before entry to main() is really pretty questionable anyway -- it allows all the threads to run **except** main, which is going to be very surprising to a typical user.

Comment (markus-becker-tridonic-com) [reply kernel/init.c:222]: > What platform?

It's a custom board based on STM32F412, but can also be seen on nrf52840-DK.

> That sounds like your logging is setup async, and doing a sleep allows a thread to run to do the output. If so... that's not a bug, it's what you asked for, technically. 

CONFIG_LOG_IMMEDIATE is indeed n, which is the default. Unsure whether I can switch it on in combination with OpenThread since that moves the logging into its context. Technically correct, still not exactly how I would have hoped

> In general, printk() isn't synchronous any more (and yes, I hate this too).

Was that an intentional change? Is there a way - that I am unaware of - to force certain prints to be immediate?

> Sleeping in the main thread before entry to main() is really pretty questionable anyway -- it allows all the threads to run **except** main, which is going to be very surprising to a typical user.

By that time of the boot process, are there any other threads than main?

Comment (jukkar) [reply kernel/init.c:222]: I would advice not using CONFIG_LOG_IMMEDIATE in any production system. I have been bit by this in networking where using the immediate option can slow things down and cause timeouts in the net stack, especially if one has various debugging options enabled.

There are networking TX & RX threads + other net related threads (depending on config) running before app main() called.

Comment (markus-becker-tridonic-com) [reply kernel/init.c:222]: Is there a printk_immediate() which I am unaware of? In cases where you want to make sure that it is immediately printed?

Comment (markus-becker-tridonic-com) [reply kernel/init.c:222]: how about using `printf()` and `k_busy_wait()` here? would that be acceptable?

**Comment by markus-becker-tridonic-com** [comment]
rebased on master post 2.3 release.

File: kernel/init.c (hunk 3 above)
Comment (markus-becker-tridonic-com) [reply kernel/init.c:222]: how about this as a solution for immediate printing:

```c
void immediate_printk(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vprintk(fmt, ap);
	va_end(ap);
}
```

```c
	if (boot_delay > 0 && IS_ENABLED(CONFIG_MULTITHREADING)) {
		immediate_printk("***** Delaying boot %u ms *****\n",
			boot_delay);
		k_busy_wait(CONFIG_BOOT_DELAY * USEC_PER_MSEC);
	}
```

Comment (markus-becker-tridonic-com) [reply kernel/init.c]: Branch updated with this proposal. Works fine for us.

**Comment by markus-becker-tridonic-com** [comment]
@jukkar @andyross would you take yet another look at this reworked solution?

File: kernel/init.c [hunk 4]
```diff
@@ -219,18 +227,18 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
 	z_stack_adjust_initialized = 1;
 #endif
 	if (boot_delay > 0 && IS_ENABLED(CONFIG_MULTITHREADING)) {
-		printk("***** delaying boot " STRINGIFY(CONFIG_BOOT_DELAY)
-		       "ms (per build configuration) *****\n");
+		immediate_printk("*** Delaying boot %u ms ***\n",
+			boot_delay);
 		k_busy_wait(CONFIG_BOOT_DELAY * USEC_PER_MSEC);
```
Comment (ceolin) [inline kernel/init.c:232]: shouldn't you use `boot_delay` here ? otherwise it will use whatever was set in CONFIG_BOOT_DELAY and not the randomized value.

File: kernel/Kconfig [hunk 5]
```diff
@@ -319,6 +319,13 @@ config BOOT_DELAY
 	  achieved by waiting for DCD on the serial port--however, not
 	  all serial ports have DCD.
 
+config BOOT_DELAY_RANDOM
+	bool "Randomize boot delay"
+	help
+	  This option randomizes the bootup delay between 0 and the
+	  value of BOOT_DELAY. This can be used e.g., to jitter boot-up
+	  of wireless network devices.
```
Comment (andyross) [inline kernel/Kconfig:327]: Pedantry: this needs to specify units.  How long is "1"?   One reasonable choice IMHO is to bake it into the API and call this CONFIG_BOOT_DELAY_RANDOM_MS

File: kernel/init.c [hunk 6]
```diff
@@ -191,6 +182,15 @@ void z_data_copy(void)
...
 bool z_sys_post_kernel;
 
+void immediate_printk(const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	vprintk(fmt, ap);
+	va_end(ap);
+}
```
Comment (andyross) [inline kernel/init.c:192]: I don't get it.  What is the point of this?

**Review by andyross — CHANGES_REQUESTED** [review]
Nitpick about docs, but one that does need to be fixed.  Also confused about that noop printk wrapper.

File: kernel/init.c [hunk 7]
```diff
@@ -191,6 +182,15 @@ void z_data_copy(void)
 
 bool z_sys_post_kernel;
 
+void immediate_printk(const char *fmt, ...)
```
Comment (ceolin) [inline kernel/init.c:185]: why are you adding it ? What it the problem with `printk` ? I presume is something related with `log_printk` ... but I don't know what it is.

File: kernel/Kconfig (hunk 5 above)
Comment (ceolin) [inline kernel/Kconfig:327]: It is worth to add a note about it in `BOOT_DELAY` option. Because there is saying that the entire value set will be used to delay the boot.

**Review by ceolin — CHANGES_REQUESTED** [review]
Please check my comments. If I'm right the random value is not being used.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
I would rather you not modify the core kernel for this use-case.

Have you considered defining a SYS_INIT() in your application (at PRE_KERNEL_1) which just introduces a delay there?

File: kernel/init.c (hunk 4 above)
Comment (markus-becker-tridonic-com) [reply kernel/init.c]: True. Corrected. As well the time units have been wrong. Corrected now as well.

File: kernel/Kconfig (hunk 5 above)
Comment (markus-becker-tridonic-com) [reply kernel/Kconfig:327]: As it says, the value is between 0 and BOOT_DELAY. BOOT_DELAY is in milliseconds:
https://github.com/zephyrproject-rtos/zephyr/pull/25627/files#diff-a0dbba97dac305f5927e75a408d79a25R309-R310

File: kernel/init.c (hunk 6 above)
Comment (markus-becker-tridonic-com) [reply kernel/init.c:192]: printk() is calling log_printk() when CONFIG_LOG_PRINTK is on. Then the output would be happening in the logging thread which is halted while the busy-loop is running. The output of the delay would only happen after the delay and thus leaving the user wondering what is happening.

This way the message is outputted before the busy-loop executes and the user knows that boot is delayed.

File: kernel/init.c (hunk 7 above)
Comment (markus-becker-tridonic-com) [reply kernel/init.c:185]: see above

File: kernel/Kconfig (hunk 5 above)
Comment (markus-becker-tridonic-com) [reply kernel/Kconfig:327]: If BOOT_DELAY_RANDOM is not set, then BOOT_DELAY is used.

Would you rather add a random delay on top of BOOT_DELAY?

**Comment by markus-becker-tridonic-com** [comment]
> I would rather you not modify the core kernel for this use-case.
> 
> Have you considered defining a SYS_INIT() in your application (at PRE_KERNEL_1) which just introduces a delay there?

Have not considered this. How would you delay network set-up with SYS_INIT()? Do you have an example?

**Comment by andyross** [comment]
> printk() is calling log_printk() when CONFIG_LOG_PRINTK is on. Then the output would be happening in the logging thread

OK... I get it.  And this is clever, but... no, sorry. :)

Really what you're doing here is **exploiting a bug** in the logging subsystem.  CONFIG_LOG_PRINTK is supposed to mean that printk() output goes to logging instead, but apparently the feature missed a spot, or put the hook in the wrong place, and it happens that vprintk() doesn't do that and goes to the original output.  But that's wrong.  When you have CONFIG_LOG=y and CONFIG_IMMEDIATE=n, logging is asynchronous.  That's the way it's supposed to work, and it's desired behavior[1].  Subsystems shouldn't be silently working around the intended behavior of other subsystems by using backdoors.

Basically: if the user asked for async logging give them what they want.  If you don't think they should want that, let's deal with that via a patch to the log layer.

[1] By someone (cough, cough, @nordic-krch).   I actually agree that for most development usage asynchronous logging is the devil's work and don't understand why we decided to make it default was a good idea.  But it's the system we have, so let's not hack holes around it.

**Comment by markus-becker-tridonic-com** [comment]
> > printk() is calling log_printk() when CONFIG_LOG_PRINTK is on. Then the output would be happening in the logging thread
> 
> OK... I get it. And this is clever

Thanks :)

> , but... no, sorry. :)

Hmm.

> Really what you're doing here is **exploiting a bug** in the logging subsystem. CONFIG_LOG_PRINTK is supposed to mean that printk() output goes to logging instead, but apparently the feature missed a spot, or put the hook in the wrong place, and it happens that vprintk() doesn't do that and goes to the original output. But that's wrong. When you have CONFIG_LOG=y and CONFIG_LOG_IMMEDIATE=n, logging is asynchronous. That's the way it's supposed to work, and it's desired behavior[1]. Subsystems shouldn't be silently working around the intended behavior of other subsystems by using backdoors.

Will the bug be fixed?

> Basically: if the user asked for async logging give them what they want. If you don't think they should want that, let's deal with that via a patch to the log layer.

True, that the user wants to use async logging (or has to because it is pulled in by default). But this tripped me several times, looking at debug on UART and expecting to see something, only to notice several trials later that the device is delaying the boot. 

For my information, would be switching the logging backend to RTT make any difference?

> [1] By someone (cough, cough, @nordic-krch). I actually agree that for most development usage asynchronous logging is the devil's work and don't understand why we decided to make it default. But it's the system we have, so let's not hack holes around it.

What was the reason to move it to async? Would disabling async logging be possible? What would be the side-effects?

**Comment by andrewboie** [comment]
> Will the bug be fixed?

Do we have a bug opened on GitHub for this issue? That's the first step.
